### ISSUE TITLE
Precisely typed IR variants

### ISSUE THREAD

**Issue by dougalm** [google-research/dex-lang#1151]
### The problem

We have several different variants of our core IR: pre-simplification, post-simplification, and various special forms used in substitutions, including IRs-within-IRs (e.g. simplified atoms within core atoms). Ideally we'd model each IR as a different Haskell type, but that would involve *lots* of boilerplate and duplicated code because they have so many constructors in common. Instead, we've modeled them all using the same Haskell types, `Atom`/`Expr`/`Block`. But it's easy to get them confused. It also means we're reluctant to add new constructors for particular use cases (e.g. a "canonical table" atom used during Imp translation) because then we'd have to handle them everywhere.

To solve this, we want a system that lets us define IR variants that share most of their constructors, but differ in some details. Here are some features we might want from such a system.
 1. Be able to traverse the IR with native Haskell case expressions
 2. Convenient injections/projections instead of many layers of constructor wrappers
 3. No performance penalty
 4. (nice to have) Open world of constructors we can keep adding to. This would be essential if we wanted to package this system as a library, but for our use case we can put up with having to list all the constructors in a central location, as long as we don't have to pattern-match on them in places where they can't occur.

### The idea

There are several ways we could tackle this. My favorite is an idea proposed by Adam which works like this: we add a phantom `IR` parameter to `Atom`, `Expr`, and other IR ADTs. `IR` is a type-level list of allowable IR fragments, e.g. `RecordIRFragment` or `LambdaIRFragment`. The constructors of `Atom`, `Expr` etc. test type-level list membership of the corresponding fragment in the `IR`. So IRs are defined as unions of IR fragments. This hits desiderata 1-3, and only falls down on 4, which we think is the least important for our use case anyway. Here's a illustration:

```haskell
data IRFragment = PairConIRF | PairProjIRF
type IR = [IRFragment]

type family HasFragment (ir::IR) (irf::IRFragment) :: Bool where
  HasFragment '[] irf = False
  HasFragment (irf  ': rest) irf = True
  HasFragment (irf' ': rest) irf = HasFragment rest irf

type HasPairCon  ir = HasFragment ir PairConIRF  ~ True
type HasPairProj ir = HasFragment ir PairProjIRF ~ True

data Expr (ir::IR) where
  Var  ::                   String             -> Expr ir
  App  ::                   Expr ir -> Expr ir -> Expr ir
  Lam  ::                   String -> Expr ir  -> Expr ir
  Pair :: HasPairCon  ir => Expr ir -> Expr ir -> Expr ir
  Fst  :: HasPairProj ir => Expr ir            -> Expr ir
  Snd  :: HasPairProj ir => Expr ir            -> Expr ir

type CoreIR = [PairConIRF, PairProjIRF] :: IR
type SimpIR = '[] :: IR

traverseSimpIR :: Expr SimpIR -> Expr SimpIR
traverseSimpIR = \case
  Var v -> Var v
  App  f x -> App f x
  Lam  b body -> Lam b body
  -- don't need Pair/Fst/Snd cases! (And GHC is smart enough to know that.)
```

### Implementation plan

I don't want this to turn into another "safer names" endless rewrite! But there's reason to think it should be much less work. Most passes will use concrete `IR` parameters, and rewriting them should just be a matter of replacing `Atom` with `Atom SimpIR` (and similar) in type signatures.

There'll be one big monolithic change where we add the `IR` parameter to everything, but without adding any constraints. It'll be a change of thousands of lines of code, but it'll only change data definitions and type signatures. Once that lands, we can incrementally make the passes more precise by adding constraints to constructors as needed. Inevitably, this will expose places where we've been sloppy about which IR we're working with (I'm especially thinking of the core vs simplified IRs here) and we'll have to tweak some passes. But the good thing is that we can do this incrementally. We can also immediately start taking advantage of the system by using it whenever we add a new constructor, like the `CanonicalTable` described earlier.

**Comment by soraros** [comment]
I believe [this](https://simon.peytonjones.org/trees-that-grow/) is relevant.

**Comment by axch** [comment]
+100 to solving this problem!  Assuming the results are readable enough, it would have helped me onboard faster, and I still feel like I'm spending rather a lot of time puzzling over which `Atom` constructors can appear where and how I should deal with them.

One desideratum I would add:
- No "incomplete pattern match" warnings.

IIUC, the proposal on the table meets this goal, but I just want to emphasize it as an explicit goal.

Finally, when I read through Trees that Grow, it felt kind of heavy-weight (not that I've actually tried it).  I hope this version with the (implicit) constraints will be more pleasant.

**Comment by dougalm** [comment]
With #1209 merged this is done!
